packages/pages/src/common/src/function/types.ts

```diff
@@ -0,0 +1,181 @@
+/**
+ * Defines the functions and fields that are available to the serverless function.
+ * @public
+ */
+export interface FunctionModule {
+  /** The exported function */
+  default?: FunctionType;
+}
+
+/**
+ * The valid Serverless Function types;
+ * @public
+ */
+export type FunctionType =
+  | HttpFunction
+  | OnPageGenerateFunction
+  | OnUrlChangeFunction;
+
+/**
+ * A function that runs when a specific path is visited on the site.
+ * @public
+ */
+export type HttpFunction = (arg: HttpFunctionArgument) => HttpFunctionResponse;
+
+/**
+ * A function that runs when a stream document is processed for HTML rendering.
+ * @public
+ */
+export type OnPageGenerateFunction = (
+  arg: OnPageGenerateArgument
+) => OnPageGenerateResponse;
+
+/**
+ * A function that runs when the path of a production page changes.
+ * @public
+ */
+export type OnUrlChangeFunction = (
+  arg: OnUrlChangeArgument
+) => OnUrlChangeResponse;
+
+/**
+ * The argument passed to a http/api type function.
+ * @public
+ */
+export interface HttpFunctionArgument {
+  /** Object containing each query parameter. */
+  queryParams: { [key: string]: string };
+  /** Object containing each path parameter. */
+  pathParams: { [key: string]: string };
+  /** Site object containing all deploy-related information. */
+  site: Site;
+}
+
+/**
+ * The return value for a http/api serverless function.
+ * @public
+ */
+export interface HttpFunctionResponse {
+  /** HTTP response body (refer to MDN Web Docs). */
+  body: string;
+  /** HTTP response status code (refer to MDN Web Docs). */
+  statusCode: number;
+  /** HTTP response headers (refer to MDN Web Docs).  */
+  headers: object;
+}
+
+/**
+ * The argument passed to an onPageGenerate type plugin.
+ * @public
+ */
+export interface OnPageGenerateArgument {
+  /** The name of the feature. */
+  feature: string;
+  /** Site object containing all deploy-related information. */
+  site: Site;
+  /** Stream object containing all stream-related information. */
+  streamOutput: any;
+}
+
+/**
+ * The return value for an onPageGenerate plugin.
+ * @public
+ */
+export interface OnPageGenerateResponse {
+  /** The path at which to serve the generate file. */
+  path: string;
+  /** The HTML content of the page */
+  content: string;
+  /** A list of optional paths at which to install redirects to the page. */
+  redirects?: string[];
+}
+
+/**
+ * The argument passed to an onUrlChange type plugin.
+ * @public
+ */
+export interface OnUrlChangeArgument {
```

Comment:
Will likely need to update after this item is done: [Slack Discussion](https://yext.slack.com/archives/C02LLE9BW2K/p1689017518031249)